src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -84,14 +84,132 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
+        {
+            return TryFormatFunctionsBlock(context, edits, source, node) ||
+                TryFormatCSharpExplicitTransition(context, edits, source, node) ||
+                TryFormatHtmlInCSharp(context, edits, source, node) ||
+                TryFormatComplexCSharpBlock(context, edits, source, node);
+        }
+
+        private static bool TryFormatFunctionsBlock(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            // @functions
+            // {
+            // }
+            if (node is CSharpCodeBlockSyntax directiveCode &&
+                directiveCode.Children.Count == 1 && directiveCode.Children.First() is RazorDirectiveSyntax directive &&
+                directive.Body is RazorDirectiveBodySyntax directiveBody &&
+                directiveBody.Keyword.GetContent().Equals("functions"))
+            {
+                var cSharpCode = directiveBody.CSharpCode;
+                var openBrace = cSharpCode.Children.FirstOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode && c.GetContent().Equals("{", StringComparison.Ordinal));
+                var closeBrace = cSharpCode.Children.LastOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode && c.GetContent().Equals("}", StringComparison.Ordinal));
+                var code = cSharpCode.Children.First(c => c.Kind == SyntaxKind.CSharpCodeBlock) as CSharpCodeBlockSyntax;
+
+                if (openBrace is null || closeBrace is null)
+                {
+                    // Don't trust ourselves in an incomplete scenario.
+                    return false;
+                }
+
+                var openBraceNode = openBrace;
+                var codeNode = code!;
+                var closeBraceNode = closeBrace;
+
+                return FormatBlock(context, source, openBraceNode, codeNode, closeBraceNode, edits, additionalIndentation: false);
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private static bool TryFormatCSharpExplicitTransition(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            // We're looking for a code block like this:
+            //
+            // @{
+            //     var x = 1;
+            // }
+            if (node is CSharpCodeBlockSyntax expliciteCode &&
+                expliciteCode.Children.First() is CSharpStatementSyntax statement &&
+                statement.Body is CSharpStatementBodySyntax csharpStatementBody)
+            {
+                var openBraceNode = csharpStatementBody.OpenBrace;
+                var codeNode = csharpStatementBody.CSharpCode;
+                var closeBraceNode = csharpStatementBody.CloseBrace;
+
+                // IndentationContext doesn't track the need for any additional indentation in this case (Both Razor and HTML desirred indentation would be "0" above).
```

Comment:
```suggestion
                // IndentationContext doesn't track the need for any additional indentation in this case (Both Razor and HTML desired indentation would be "0" above).
```

Reply:
This comment doesn't parse for me, I'm unsure what the context is for desired indentation being 0, what _above_ is.

Reply:
After reading your comment [here](https://github.com/dotnet/aspnetcore-tooling/pull/4032#discussion_r685515677) I had all the same concerns @davidwengier had. I wonder if the reason why this is unique is because it's special cased at the indentation context creation layer?

Something we need to understand because it represents a point of failure for future changes. Not saying you need to investigate it now but we should have a fundamentals issue to better understand why this is happening only for this case.

Reply:
Good eye being concerned about this. I did some deeper digging and it turned out that `context.Indentations[lineNumber]` was the wrong thing to do. The `FormattingVisitor` tracks things on a more granular level than that, so you wanna give the absolutePosition of the point you're trying to format and it will give you the correct indentation level.